### ISSUE TITLE
polars-learn

### ISSUE THREAD

**Issue by firmai** [abstractqqq/polars_ds_extension#202]
Always funny how these projects start, one goes from like 100 users that understand the tremendous opportunity of fast in-memory computing --- and then 2-3 years later 10 million people heavily rely on your solutions. 

There is a lot of potential in your project, it lacks a name, clarity (too much text on readme), marketing, and objectives. I think polars extensions and libraries built ontop of polars is the future of data science for the next 10 years. 

Similar to rust-ml/linfa, I think you should aim for something much more grand. Why not search active open-source sponsorship, corporate or otherwise. 

There is currently no polars sklearn. **How could this be?**. It is currently the most obvious missing link in data science. It is for a DS geek a wheeled suitcase moment - "We put wheels on bags after we put a man on the moon". 

We need clustering, we need dimensionality reduction, feature selection, pairwise interactions, we need a polars machine learning and data-science project worth throwing yourself behind. Everybody is so busy creating LLMs to generate chicken casserol recipes, nobody is doing any actual work. 

What are the hundreds and thousands of data scientists doing in their free time? Thanks for picking up the slack, you will be tomorrow's heros.

**Comment by firmai** [comment]
Hopefully you have seen this solution https://github.com/azmyrajab/polars_ols

You should get your company onboard OS packages are becoming the best marketing for great talent acquisition

**Comment by abstractqqq** [comment]
Yep I am fully aware of polars_ols. The main thing is that I do not want to introduce dependency on third party blas or Lapack distributions.. It's a nightmare configuring all that. I am betting on Faer-rs, which is an alternative to those old C/Fortran linear algebra libraries. I take dependency very seriously. You can read up on how SciPy almost couldn't compile for Python 3.12 because of old Fortran dependency.  So far for linear regression and SVD, speed is on par and even better in some cases and the author seems to be very knowledgeable.

**Comment by abstractqqq** [comment]
@firmai  By the way, I am an NYU alumni :)

**Comment by firmai** [comment]
functime developers don't seem that interested in the pacakge, they still haven't updated for polars 1, don't you think it is better to asborb it into your package https://github.com/functime-org/functime/issues/250

**Comment by abstractqqq** [comment]
> functime developers don't seem that interested in the pacakge, they still haven't updated for polars 1, don't you think it is better to asborb it into your package https://github.com/functime-org/functime/issues/250
> 
> 

Yes and no. I have been adding tsfresh style features slowly. Currently they are scattered around in num.py and stats.py and I haven't consolidated the features. Functime did a huge project of rewriting most Tsfresh features and I was part of the project, and I did a lot of performance testing and wrote more optimal queries for a lot of the features... So yes, I can take care of the feature extraction easily. I have more than half of what tsfresh and functime offers. And yes again the low recognition is because I am not actively marketing.. Sigh...

Although I like functional programming, I find it hard to track states in Functimes's transforms and I am increasingly feeling that classes are fine as long as they are shallow and serve one focus.. 

Time series transforms can be very different from traditional tabular ML transforms and 
I do not do time series a lot in my work and I am mostly learning along the way so it will take some time for me. I can go and ask the maintainers about the future of functime..

**Comment by firmai** [comment]
> Yep I am fully aware of polars_ols. The main thing is that I do not want to introduce dependency on third party blas or Lapack distributions.. It's a nightmare configuring all that. I am betting on Faer-rs, which is an alternative to those old C/Fortran linear algebra libraries. I take dependency very seriously. You can read up on how SciPy almost couldn't compile for Python 3.12 because of old Fortran dependency. So far for linear regression and SVD, speed is on par and even better in some cases and the author seems to be very knowledgeable.

how is your OLS coming along?

**Comment by abstractqqq** [comment]
Ridge, Lasso, Rolling and Recursive are added in v0.5.1 (a bugged version of Ridge was introduced in v0.5.0). I have made some changes and improvements to all of these since the release. You can find them in the docs here:

https://polars-ds-extension.readthedocs.io/en/latest/num.html#polars_ds.num.query_rolling_lstsq

More null_policies will be good. But that can be tricky... 

We also developed some benchmarks vs. sklearn. These are not strictly apples-to-apples, because the default "solver" may not be the same. For Lasso, I am also "not minding the dual gap" at this moment because I do not understand it well enough. (Also practically, I think it is enough to stop coordinate descent when the updates are small). Anyways, using the default, we have some good numbers:  

![image](https://github.com/user-attachments/assets/2924c417-0db6-4bc3-af05-0dcb120f9741)

**Comment by abstractqqq** [comment]
Turns out standalone linear regression is EASY. A regular LR class (linear regression) and an Online LR class has been implemented. 

I also added weighted lstsq as an option in query_lstsq, and linear regression with rcond.

A new ver will be released this coming weekend and I would like to take a break from linear regression.. Next is likely k-means, standalone kdtree and the ball-tree algorithm.

**Comment by fingoldo** [comment]
> What are the hundreds and thousands of data scientists doing in their free time?

I keep asking myself the same question every day )
